### PR TITLE
[1ES] Add 1ES pipeline files

### PR DESCRIPTION
**Pull request by liliankasem**
### Issue describing the changes in this PR

resolves #issue_for_this_pr

### Pull request checklist

* [x] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
  * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

This PR introduces new CI files for a new build pipeline.

### CODE REVIEW

File: eng/ci/code-mirror.yml [hunk 1]
```diff
@@ -4,6 +4,7 @@ trigger:
     # Keep this set limited as appropriate (don't mirror individual user branches).
     - main
     - release/*
+    - feature/*
```
Comment (jviau) [inline eng/ci/code-mirror.yml:7]: Do we need official builds of feature branches? If so, we need to also lock down creation of these branches.

If we just want CI validation of feature branches, we can look into using unofficial 1ES PT builds for that in our public org.

File: eng/ci/templates/jobs/run-integration-tests-linux.yml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  - task: DotNetCoreCLI@2
+    displayName: Build DotNetWorker
+    inputs:
+      command: build
+      arguments: '-c Release'
+      projects: DotNetWorker.sln
+
+  - pwsh: |
+      $useIntegrationTestingCoreTools = $null
+      if (-not ([bool]::TryParse($env:USECORETOOLSBUILDFROMINTEGRATIONTESTS, [ref] $useIntegrationTestingCoreTools)))
```
Comment (jviau) [inline eng/ci/templates/jobs/run-integration-tests-linux.yml:25]: Would a template parameter be more fitting for this? You can restrict it to `boolean` with a default value. Can then switch to PowerShellTask and invoke `./setup-e2e-tests.ps1` directly with args

File: eng/ci/templates/jobs/run-integration-tests-linux.yml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      {
+        throw "UseCoreToolsBuildFromIntegrationTests can only be set to True or False. Current value is set to $env:USECORETOOLSBUILDFROMINTEGRATIONTESTS"
+      }
+      ./setup-e2e-tests.ps1 -DotnetVersion 'net7' `
+                            -UseCoreToolsBuildFromIntegrationTests:$useIntegrationTestingCoreTools `
+                            -SkipCosmosDBEmulator `
+                            -SkipBuildOnPack
+    displayName: "Setup E2E tests"
+    env:
+      CORE_TOOLS_URL: $(CORE_TOOLS_URL)
```
Comment (jviau) [inline eng/ci/templates/jobs/run-integration-tests-linux.yml:35]: Is this needed? Powershell will already be inheriting that environment variable.

File: eng/ci/templates/jobs/run-integration-tests-windows.yml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      if (-not ([bool]::TryParse($env:USECORETOOLSBUILDFROMINTEGRATIONTESTS, [ref] $useIntegrationTestingCoreTools)))
+      {
+        throw "UseCoreToolsBuildFromIntegrationTests can only be set to True or False. Current value is set to $env:USECORETOOLSBUILDFROMINTEGRATIONTESTS"
+      }
+      ./setup-e2e-tests.ps1 -DotnetVersion 'net7' `
+                            -UseCoreToolsBuildFromIntegrationTests:$useIntegrationTestingCoreTools `
+                            -SkipBuildOnPack
+    displayName: '[NET 7] Setup E2E tests'
+
+  - task: DotNetCoreCLI@2
```
Comment (jviau) [inline eng/ci/templates/jobs/run-integration-tests-windows.yml:37]: Consider pulling framework into a template parameter, then call this template twice: once for net7, once for netfx. (And then deleting the duplicate pwsh and test steps)

File: eng/ci/templates/jobs/run-unit-tests-windows.yml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+- job: RunUnitTests_Windows
+  displayName: '[Windows] Run Unit Tests'
+
+  pool:
+    name: 1es-pool-azfunc
+    image: 1es-windows-2022
+    os: windows
+
+  steps:
+  - task: NuGetAuthenticate@1
```
Comment (jviau) [inline eng/ci/templates/jobs/run-unit-tests-windows.yml:11]: Did you copy this from the host CI files? I meant to remove that, it is leftover from some workaround I had earlier. You shouldn't need this step anywhere, we don't use any authenticated nuget feeds.

File: eng/ci/templates/official/jobs/build-artifacts.yml [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+      path: $(Build.ArtifactStagingDirectory)/NugetPackages
+      artifact: NugetPackages
+
+  pool:
+    name: 1es-pool-azfunc
+    image: 1es-windows-2022
+    os: windows
+
+  variables:
+    buildNumber: $(Build.BuildNumber)
```
Comment (jviau) [inline eng/ci/templates/official/jobs/build-artifacts.yml:19]: You will need to copy the build number logic from our existing CI files, as we want this to be empty string on release builds.

File: eng/ci/templates/official/jobs/build-artifacts.yml [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    displayName: Delete CodeSignSummary files
+    inputs:
+      contents: '**/CodeSignSummary-*.md'
+
+  - task: DotNetCoreCLI@2
+    displayName: Pack SDK
+    inputs:
+      command: custom
+      custom: pack
+      arguments: '--no-build -c release -p:PackageOutputPath=$(System.DefaultWorkingDirectory)\out\pkg\release -p:BuildNumber=$(buildNumber)'
```
Comment (jviau) [inline eng/ci/templates/official/jobs/build-artifacts.yml:146]: I think you can use:

```suggestion
      arguments: '--no-build -c release -o $(System.DefaultWorkingDirectory)\out\pkg\release -p:BuildNumber=$(buildNumber)'
```

Also, no need to copy `out\pkg` style here. You could pack directly to `$(Build.ArtifactStagingDirectory)/NugetPackages` and then sign inline there also.

File: eng/ci/templates/jobs/run-unit-tests-linux.yml [hunk 8]
```diff
@@ -0,0 +1,16 @@
+jobs:
+- job: RunUnitTests_Linux
```
Comment (jviau) [inline eng/ci/templates/jobs/run-unit-tests-linux.yml:2]: You can remove the separate files for windows and linux and include pool as a parameter:

``` yml
parameters:
  - name: os
     type: string
     values:
       - windows
       - linux
.
.
.
  - ${{ if eq(parameters.os, 'windows') }}:
    pool:
      name: 1es-pool-azfunc
      image: 1es-windows-2022
      os: windows
  - ${{ elseif eq(parameters.os, 'linux') }}:
     pool:
       name: 1es-pool-azfunc
       image: 1es-ubuntu-22.04
       os: linux
```

Can also us `${{ parameters.os }}` expression as part of the display name and job name

Comment (jviau) [reply eng/ci/templates/jobs/run-unit-tests-linux.yml:2]: Or maybe strategy/matric will work here also:
https://learn.microsoft.com/en-us/azure/devops/pipelines/yaml-schema/jobs-job-strategy?view=azure-pipelines

File: eng/ci/templates/jobs/run-integration-tests-windows.yml (hunk 4 above)
Comment (jviau) [reply eng/ci/templates/jobs/run-integration-tests-windows.yml:37]: Or maybe strategy/matric will work here also:
https://learn.microsoft.com/en-us/azure/devops/pipelines/yaml-schema/jobs-job-strategy?view=azure-pipelines

File: eng/ci/code-mirror.yml (hunk 1 above)
Comment (liliankasem) [reply eng/ci/code-mirror.yml:7]: That's a good point, we probably don't need that and can just use the unofficial build

File: eng/ci/templates/jobs/run-integration-tests-linux.yml (hunk 3 above)
Comment (liliankasem) [reply eng/ci/templates/jobs/run-integration-tests-linux.yml:35]: Will remove it and see what happens

File: eng/ci/templates/jobs/run-unit-tests-linux.yml (hunk 8 above)
Comment (liliankasem) [reply eng/ci/templates/jobs/run-unit-tests-linux.yml:2]: Oooo I like this idea, will do

File: eng/ci/templates/jobs/run-unit-tests-windows.yml (hunk 5 above)
Comment (liliankasem) [reply eng/ci/templates/jobs/run-unit-tests-windows.yml:11]: guilty

Comment (jviau) [reply eng/ci/templates/jobs/run-unit-tests-windows.yml:11]: It was helpful though, reminded me to go clean that up myself lol

**Review by jviau — APPROVED** [review]

**Pull request merged**